File: apps/web/lib/permissions.ts
/** Mirror of backend RBAC — keep in sync with apps/api/exposureflow_api/auth/permissions.py */

export type WorkspaceRole =
  | "owner"
  | "admin"
  | "strategist"
  | "editor"
  | "analyst"
  | "client_viewer"
  | "billing_admin"
  | "support_admin";

export const ROLE_LABELS: Record<WorkspaceRole, string> = {
  owner: "擁有者",
  admin: "管理員",
  strategist: "策略師",
  editor: "編輯",
  analyst: "分析師",
  client_viewer: "客戶檢視",
  billing_admin: "計費管理",
  support_admin: "平台支援",
};

const PERMISSIONS: Record<WorkspaceRole, Set<string>> = {
  owner: new Set([
    "workspace:read",
    "workspace:write",
    "site:read",
    "site:write",
    "member:read",
    "member:write",
    "invitation:write",
    "integration:read",
    "integration:write",
    "job:read",
    "job:write",
    "api_key:write",
    "billing:read",
    "agency:read",
    "ops:read",
    "impersonate",
    "client:approve",
  ]),
  admin: new Set([
    "workspace:read",
    "workspace:write",
    "site:read",
    "site:write",
    "member:read",
    "member:write",
    "invitation:write",
    "integration:read",
    "integration:write",
    "job:read",
    "job:write",
    "api_key:write",
    "billing:read",
    "agency:read",
    "ops:read",
    "client:approve",
  ]),
  client_viewer: new Set(["workspace:read", "site:read", "client:approve"]),
  strategist: new Set([
    "workspace:read",
    "site:read",
    "site:write",
    "member:read",
    "integration:read",
    "job:read",
    "job:write",
    "client:approve",
  ]),
  editor: new Set([
    "workspace:read",
    "site:read",
    "site:write",
    "job:read",
    "client:approve",
  ]),
  analyst: new Set(["workspace:read", "site:read", "job:read"]),
  billing_admin: new Set(["workspace:read", "billing:read"]),
  support_admin: new Set(["workspace:read", "site:read", "member:read", "impersonate"]),
};

export function canRole(role: string | undefined, permission: string): boolean {
  if (!role) return false;
  return PERMISSIONS[role as WorkspaceRole]?.has(permission) ?? false;
}

export function isPlatformSupport(role: string | undefined): boolean {
  return role === "support_admin";
}

export function usesClientPortal(role: string | undefined): boolean {
  return role === "client_viewer";
}

export function usesBillingOnlyShell(role: string | undefined): boolean {
  return role === "billing_admin";
}

export function resolveEntryPath(
  workspaceId: string,
  role: string | undefined,
  siteId?: string,
): string {
  if (usesClientPortal(role)) return `/client/${workspaceId}`;
  if (usesBillingOnlyShell(role)) return `/app/${workspaceId}/settings/billing`;
  if (isPlatformSupport(role)) return `/internal-admin/workspaces`;
  if (siteId) return `/app/${workspaceId}/sites/${siteId}/dashboard`;
  return `/app/${workspaceId}/onboarding`;
}

File: apps/api/exposureflow_api/auth/permissions.py
"""RBAC permission matrix for workspace-scoped operations."""

from collections.abc import Callable
from typing import Any
from uuid import UUID

from fastapi import Depends
from sqlalchemy.ext.asyncio import AsyncSession

from exposureflow_api.auth.deps import get_current_user, get_workspace_membership, require_workspace_access
from exposureflow_api.auth.jwt import AuthContext
from exposureflow_api.common.errors import APIError
from exposureflow_api.database import get_db
from exposureflow_api.models import WorkspaceMembership

ROLE_RANK: dict[str, int] = {
    "client_viewer": 10,
    "analyst": 20,
    "editor": 30,
    "strategist": 40,
    "billing_admin": 45,
    "admin": 50,
    "owner": 60,
    "support_admin": 70,
}

PERMISSIONS: dict[str, set[str]] = {
    "owner": {
        "workspace:read",
        "workspace:write",
        "site:read",
        "site:write",
        "member:read",
        "member:write",
        "invitation:write",
        "integration:read",
        "integration:write",
        "job:read",
        "job:write",
        "api_key:write",
        "billing:read",
        "impersonate",
        "client:approve",
    },
    "admin": {
        "workspace:read",
        "workspace:write",
        "site:read",
        "site:write",
        "member:read",
        "member:write",
        "invitation:write",
        "integration:read",
        "integration:write",
        "job:read",
        "job:write",
        "api_key:write",
        "billing:read",
        "client:approve",
    },
    "client_viewer": {
        "workspace:read",
        "site:read",
        "client:approve",
    },
    "strategist": {
        "workspace:read",
        "site:read",
        "site:write",
        "member:read",
        "integration:read",
        "job:read",
        "job:write",
        "client:approve",
    },
    "editor": {
        "workspace:read",
        "site:read",
        "site:write",
        "job:read",
        "client:approve",
    },
    "analyst": {
        "workspace:read",
        "site:read",
        "job:read",
    },
    "billing_admin": {
        "workspace:read",
        "billing:read",
    },
    "support_admin": {
        "workspace:read",
        "site:read",
        "member:read",
        "impersonate",
    },
}


def role_has_permission(role: str, permission: str) -> bool:
    return permission in PERMISSIONS.get(role, set())


def require_permission(permission: str) -> Callable[..., Any]:
    async def _check(
        ctx: tuple[AuthContext, WorkspaceMembership, UUID] = Depends(require_workspace_access),
    ) -> tuple[AuthContext, WorkspaceMembership, UUID]:
        _user, membership, workspace_id = ctx
        if not role_has_permission(membership.role, permission):
            raise APIError(
                code="PERMISSION_DENIED",
                message=f"Role '{membership.role}' cannot perform '{permission}'.",
                status_code=403,
            )
        return ctx

    return _check


def require_min_role(min_role: str) -> Callable[..., Any]:
    min_rank = ROLE_RANK[min_role]

    async def _check(
        workspace_id: UUID,
        user: AuthContext = Depends(get_current_user),
        db: AsyncSession = Depends(get_db),
    ) -> WorkspaceMembership:
        membership = await get_workspace_membership(workspace_id, user, db)
        if ROLE_RANK.get(membership.role, 0) < min_rank:
            raise APIError(
                code="PERMISSION_DENIED",
                message=f"Minimum role '{min_role}' required.",
                status_code=403,
            )
        return membership

    return _check
